File: typescript/parser.ts
/// <reference path="maths.ts" />

module Parser {
    function parseHelper(tokens: string[], i: number) : {
      end: number, 
      e : Maths.Expression 
    } {
      var s = tokens[i];

      if ("-+*/^%".indexOf(s) >= 0) {
        var left = parseHelper(tokens, i + 1);
        var right = parseHelper(tokens, left.end);

        var ops = { 
          '+': Maths.Sum,
          '-': Maths.Difference,
          '*': Maths.Product,
          '/': Maths.Quotient,
          '%': Maths.Modulo,
          '^': Maths.Power
        };
        var op = ops[s];
        if (op === undefined) { throw "op '" + s + "' is not defined"; }

        var e: Maths.Expression = new op(left.e, right.e);
        return { end: right.end, e };
      }

      return { end: i + 1, e: new Maths.Constant(+s) };
  }

  export function parse(s: string) : Maths.Expression {
      var regex = /-?\d+|[-+*/^%]/g;

      var tokens: string[] = [];
      var m;
      while (m = regex.exec(s)) {
        tokens.push(m[0]);
      }

      var result = parseHelper(tokens, 0);
      if (result.end !== tokens.length) {
        throw "enexpected EOF";
      }
      return result.e;
  }
}

File: csharp/Parser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace Math
{
    static class Parser
    {
        static Regex lexer = new Regex(@"-?\d+|[-+*/^%]", RegexOptions.Compiled);
        public static Expression Parse(string s) {
            var tokens = Lex(s).ToList();
            var (e, end) = ParseHelper(tokens, 0);
            if (end != tokens.Count) {
                throw new InvalidOperationException("Expected EOF");
            }
            return e;
        }

        private static IEnumerable<string> Lex(string s) {
            foreach (Match match in lexer.Matches(s)) {
                yield return match.Value;
            }
        }

        private static (Expression, int) ParseHelper(List<string> tokens, int start) {
            var s = tokens[start];

            if("-+*/^%".Contains(s)) {
                var (x, next) = ParseHelper(tokens, start + 1);
                var (y, end) = ParseHelper(tokens, next);

                switch(s) {
                    case "+": return (new Sum(x, y), end);
                    case "-": return (new Difference(x, y), end);
                    case "*": return (new Product(x, y), end);
                    case "/": return (new Quotient(x, y), end);
                    case "^": return (new Power(x, y), end);
                    case "%": return (new Modulo(x, y), end);
                    default: throw new InvalidOperationException($"Unknown op {s}");
                }
            }

            if (int.TryParse(s, out var n)) {
                return (new Constant { Value = n }, start + 1);
            }

            throw new InvalidOperationException($"Unknown at: {s} #{start}");
        }
    }
}